Comorphisms/CASL2Skolem.hs

```diff
@@ -0,0 +1,200 @@
+{-# LANGUAGE MultiParamTypeClasses, TypeSynonymInstances, FlexibleInstances #-}
+{- |
+Module      :  $Header$
+Description :  skolemization as an institution comorphism
+Copyright   :  (c) Mihai Codescu, 2016
+License     :  GPLv2 or higher, see LICENSE.txt
+
+Maintainer  :  codescu@iws.cs.uni-magdeburg.de
+Stability   :  provisional
+Portability :  non-portable (imports Logic.Comorphism)
+
+removing existential quantifiers from every formula
+follows http://resources.mpi-inf.mpg.de/departments/rg1/teaching/autrea-ss10/script/lecture10.pdf
+
+-}
+
+module Comorphisms.CASL2Skolem where
+
+import Logic.Logic
+import Logic.Comorphism
+
+import CASL.Logic_CASL
+import CASL.AS_Basic_CASL
+import CASL.Sign
+import CASL.Morphism
+import CASL.Sublogic as SL hiding (bottom)
+import CASL.Induction
+import CASL.Quantification
+
+import Common.Result
+import Common.Id
+import qualified Data.Set as Set
+import Common.AS_Annotation
+import Common.ProofTree
+import qualified Common.Lib.MapSet as MapSet
+
+import qualified Common.Lib.Rel as Rel
+
+import Data.List(partition)
+
+data CASL2Skolem = CASL2Skolem deriving Show
+
+instance Language CASL2Skolem where
+    language_name CASL2Skolem = "CASL2Skolem"
+
+instance Comorphism CASL2Skolem
+               CASL CASL_Sublogics
+               CASLBasicSpec CASLFORMULA SYMB_ITEMS SYMB_MAP_ITEMS
+               CASLSign
+               CASLMor
+               Symbol RawSymbol ProofTree
+               CASL CASL_Sublogics
+               CASLBasicSpec CASLFORMULA SYMB_ITEMS SYMB_MAP_ITEMS
+               CASLSign
+               CASLMor
+               Symbol RawSymbol ProofTree where
+    sourceLogic CASL2Skolem = CASL
+    sourceSublogic CASL2Skolem = SL.caslTop --TODO
+    targetLogic CASL2Skolem = CASL
+    mapSublogic CASL2Skolem sl = Just sl -- TODO
+    map_theory CASL2Skolem = mapTheory
+    map_morphism CASL2Skolem = error "nyi"
+    map_sentence CASL2Skolem _ _ = error "nyi"
+    map_symbol CASL2Skolem _ = Set.singleton . id
+    has_model_expansion CASL2Skolem = True -- check
+    is_weakly_amalgamable CASL2Skolem = True --check
```

Comment:
again, since model translation is the identity, both properties hold. But we have a severe problem: the satisfaction condition does not hold. Indeed, the Skolem transformation does not preserve satisfaction in a given model, but only satisfiability. Proposed solution: introduce two more flags, namely eps and rps (expanstion resp. reduction preserves satisfaction, see the terminology of http://matwbn.icm.edu.pl/ksiazki/bcp/bcp28/bcp2815.pdf , there used only for the satisfaction condition of institutiosn). Normally, a comorphism has eps and rps both true, but this one has rps true and eps false, because e.g. M |= alpha(\exists x. P(x)) <=> M |= P(a) => \beta(M)=M|=\exists x. P(x), but not vice versa (\beta plays the role of "reduction" here).